<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from robot_lqr.xacro                | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="robot_wbr">
  <!-- 宏:黑色设置 -->
  <material name="black">
    <color rgba="0.0 0.0 0.0 1.0"/>
  </material>
  <!-- 质量  -->
  <!-- 质量  -->
  <!-- 底盘 -->
  <link name="base_footprint">
    <visual>
      <geometry>
        <sphere radius="0.001"/>
      </geometry>
    </visual>
  </link>
  <link name="base_link">
    <visual>
      <geometry>
        <box size="0.02 0.5 0.02"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <material name="yellow">
        <color rgba="0.5 0.3 0.0 0.5"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.02 0.5 0.02"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </collision>
    <!-- <inertial> -->
    <mass value="0.01"/>
    <inertia ixx="0.00020866666666666668" ixy="0" ixz="0" iyy="6.666666666666668e-07" iyz="0" izz="0.00020866666666666668"/>
    <!-- </inertial> -->
  </link>
  <joint name="base_link2base_footprint" type="fixed">
    <parent link="base_footprint"/>
    <child link="base_link"/>
    <origin xyz="0 0 0.05"/>
  </joint>
  <gazebo reference="base_link">
    <material>Gazebo/Yellow</material>
  </gazebo>
  <link name="vertical_link">
    <visual>
      <geometry>
        <box size="0.02 0.02 0.5"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <material name="yellow">
        <color rgba="0.5 0.3 0.0 0.5"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.02 0.02 0.5"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </collision>
    <inertia>
      <!-- <inertial> -->
      <mass value="0.5"/>
      <inertia ixx="0.010433333333333334" ixy="0" ixz="0" iyy="0.010433333333333334" iyz="0" izz="3.3333333333333335e-05"/>
      <!-- </inertial> -->
    </inertia>
  </link>
  <joint name="base2vertial_link" type="fixed">
    <parent link="base_link"/>
    <child link="vertical_link"/>
    <origin xyz="0 0 0.25"/>
  </joint>
  <gazebo reference="vertical_link">
    <material>Gazebo/Yellow</material>
  </gazebo>
  <link name="left_wheel">
    <visual>
      <geometry>
        <cylinder length="0.02" radius="0.05"/>
      </geometry>
      <origin rpy="1.5707963 0.0 0.0" xyz="0.0 0.0 0.0"/>
      <material name="black"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.02" radius="0.05"/>
      </geometry>
      <origin rpy="1.5707963 0.0 0.0" xyz="0.0 0.0 0.0"/>
    </collision>
    <inertial>
      <mass value="0.8"/>
      <inertia ixx="0.0005266666666666667" ixy="0" ixz="0" iyy="0.0005266666666666667" iyz="0" izz="0.0010000000000000002"/>
    </inertial>
  </link>
  <joint name="left_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="left_wheel"/>
    <origin xyz="0 0.25 0 "/>
    <axis xyz="0 1 0"/>
  </joint>
  <gazebo reference="left_wheel">
    <material>Gazebo/Black</material>
  </gazebo>
  <link name="right_wheel">
    <visual>
      <geometry>
        <cylinder length="0.02" radius="0.05"/>
      </geometry>
      <origin rpy="1.5707963 0.0 0.0" xyz="0.0 0.0 0.0"/>
      <material name="black"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.02" radius="0.05"/>
      </geometry>
      <origin rpy="1.5707963 0.0 0.0" xyz="0.0 0.0 0.0"/>
    </collision>
    <inertial>
      <mass value="0.8"/>
      <inertia ixx="0.0005266666666666667" ixy="0" ixz="0" iyy="0.0005266666666666667" iyz="0" izz="0.0010000000000000002"/>
    </inertial>
  </link>
  <joint name="right_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="right_wheel"/>
    <origin xyz="0 -0.25 0 "/>
    <axis xyz="0 1 0"/>
  </joint>
  <gazebo reference="right_wheel">
    <material>Gazebo/Black</material>
  </gazebo>
  <link name="mass_link">
    <visual>
      <geometry>
        <box size="0.3 0.3 0.2"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <material name="yellow">
        <color rgba="0.5 0.3 0.0 0.5"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.3 0.3 0.2"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </collision>
    <inertia>
      <!-- <inertial> -->
      <mass value="10"/>
      <inertia ixx="0.10833333333333334" ixy="0" ixz="0" iyy="0.10833333333333334" iyz="0" izz="0.15"/>
      <!-- </inertial> -->
    </inertia>
  </link>
  <joint name="mass2vertial_joint" type="revolute">
    <parent link="vertical_link"/>
    <child link="mass_link"/>
    <origin xyz="0 0 0.25"/>
    <axis xyz="0 1 0"/>
    <limit effort="50" lower="-1.57" upper="1.57" velocity="20.0"/>
  </joint>
  <gazebo reference="mass_link">
    <material>Gazebo/Red</material>
  </gazebo>
  <!-- 调用机器人模型 -->
  <transmission name="right_wheel_joint_transmission">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="right_wheel_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <!-- 在 Gazebo      当中用 ： EffortJointInterface -->
      <!-- 在 RobotHW     当中用 ： hardware_interface/EffortJointInterface -->
    </joint>
    <actuator name="right_wheel_joint_motor">
      <mechanicalReduction>1</mechanicalReduction>
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <!-- 在 Indigo 之前的版本，hardwareInterface只在此处定义（和我noetic什么关系x） -->
    </actuator>
  </transmission>
  <transmission name="left_wheel_joint_transmission">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="left_wheel_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <!-- 在 Gazebo      当中用 ： EffortJointInterface -->
      <!-- 在 RobotHW     当中用 ： hardware_interface/EffortJointInterface -->
    </joint>
    <actuator name="left_wheel_joint_motor">
      <mechanicalReduction>1</mechanicalReduction>
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <!-- 在 Indigo 之前的版本，hardwareInterface只在此处定义（和我noetic什么关系x） -->
    </actuator>
  </transmission>
  <transmission name="mass2vertial_joint_transmission">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="mass2vertial_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <!-- 在 Gazebo      当中用 ： EffortJointInterface -->
      <!-- 在 RobotHW     当中用 ： hardware_interface/EffortJointInterface -->
    </joint>
    <actuator name="mass2vertial_joint_motor">
      <mechanicalReduction>1</mechanicalReduction>
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <!-- 在 Indigo 之前的版本，hardwareInterface只在此处定义（和我noetic什么关系x） -->
    </actuator>
  </transmission>
  <!-- 启动gazebo_ros_control 插件，命名空间设置为/robot_wbr -->
  <gazebo>
    <plugin filename="libgazebo_ros_control.so" name="gazebo_ros_control">
      <robotNamespace>/robot_lqr2</robotNamespace>
    </plugin>
  </gazebo>
  <!-- 绑定机器人驱动器（注意关节类型）+ 开启gazebo_ros_control 插件（注意命名空间）-->
  <!-- 由于imu与body是固连的，所以直接绑base_link ok-->
  <gazebo reference="mass_link">
    <gravity>true</gravity>
    <sensor name="imu_sensor" type="imu">
      <always_on>true</always_on>
      <update_rate>100</update_rate>
      <visualize>true</visualize>
      <topic>__default_topic__</topic>
      <plugin filename="libgazebo_ros_imu_sensor.so" name="imu_plugin">
        <topicName>imu</topicName>
        <bodyName>imu_link</bodyName>
        <!-- 此处也有一个bodyName -->
        <updateRateHZ>100.0</updateRateHZ>
        <gaussianNoise>0.0</gaussianNoise>
        <xyzOffset>0 0 0</xyzOffset>
        <rpyOffset>0 0 0</rpyOffset>
        <frameName>imu_link</frameName>
        <!-- 这个frameName是否可以使用imu_link 需要经过测试，因为原本这个代码的参考bodyName并非imu_link，这是我起的link名字 -->
      </plugin>
      <pose>0 0 0 0 0 0</pose>
    </sensor>
  </gazebo>
  <!-- 绑定外部传感器（内部传感器可以通过机器人状态读取） -->
</robot>

